### PR TITLE
GH-39217: [Python] RecordBatchReader.from_stream constructor for objects implementing the Arrow PyCapsule protocol

### LINKED ISSUE #39217: [Python] RecordBatchReader constructor from stream object implementing the PyCapsule Protocol
In https://github.com/apache/arrow/pull/37797 we added the dunder methods for the Arrow PyCapsule Protocol, and we also already added support for checking for objects that implement the protocol in the `pa.array(..)`, `pa.record_batch(..)` and `pa.schema(..)` constructors, such that you can for example create a pyarrow array with `pa.array(obj)` given any object `obj` that supports the interface by defining `__arrow_c_array__`.

But for the stream objects, we don't have an equivalent factory function that creates a RecordBatchReader. Therefore I think it would be good to add a public RecordBatchReader constructor from stream objects implementing the protocol (to avoid you need to call the `_import_from_c_capsule` private method for this use case). For example `RecordBatchReader.from_stream`?

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
### Rationale for this change

In contrast to Array, RecordBatch and Schema, for the C Stream (mapping to RecordBatchReader) we haven't an equivalent factory function that can accept any Arrow-compatible object and turn it into a pyarrow object through the PyCapsule Protocol.

For that reason, this proposes an explicit constructor class method for this: `RecordBatchReader.from_stream` (this is a quite generic name, so other name suggestions are certainly welcome).

### Are these changes tested?
TODO

* Closes: #39217

### CODE REVIEW

File: python/pyarrow/ipc.pxi [hunk 1]
```diff
@@ -883,6 +883,43 @@ cdef class RecordBatchReader(_Weakrefable):
...
+
+        if not hasattr(data, "__arrow_c_stream__"):
+            raise TypeError(
+                "Expected an object implementing the Arrow PyCapsule Protocol for "
+                "streams (i.e. having a `__arrow_c_stream__` method), "
+                f"got {type(data)!r}."
+            )
+
+        if schema is not None:
+            requested = schema.__arrow_c_schema__()
```
Comment (pitrou) [inline python/pyarrow/ipc.pxi:916]: Do we also want to first test the presence of this method using `hasattr` as above?

**Review by pitrou — COMMENTED** [review]
This is a good idea. Of course it needs some tests.
Should `pyarrow.ipc.open_stream` also accept PyCapsule producers?

File: python/pyarrow/ipc.pxi (hunk 1 above)
Comment (jorisvandenbossche) [reply python/pyarrow/ipc.pxi:916]: Yes, good idea

**Comment by jorisvandenbossche** [comment]
> Should `pyarrow.ipc.open_stream` also accept PyCapsule producers?

My understanding is that at the moment this function is meant to work with file-like objects (or the in-memory buffer representing it) for an IPC encapsulated message. That seems a bit different in scope, and I would say it's fine to keep that scope?

**Comment by pitrou** [comment]
> My understanding is that at the moment this function is meant to work with file-like objects (or the in-memory buffer representing it) for an IPC encapsulated message. That seems a bit different in scope, and I would say it's fine to keep that scope?

Fair enough. My concern was that `RecordBatchReader.from_stream` is less visible...

**Review by pitrou — APPROVED** [review]

**Pull request merged**